internal/controller/system/controller.go

```diff
@@ -0,0 +1,181 @@
+					return time.Duration(i+1) * ctrl.databaseRetryConfiguration.Delay
+				}
+
+				return 0
+			}),
+		)
+
+		// Add cache regarding database state
+		ledgerController = ledgercontroller.NewControllerWithCache(*l, ledgerController, ctrl.registry)
+
+		// Add traces
+		ledgerController = ledgercontroller.NewControllerWithTraces(ledgerController, ctrl.tracer)
+
+		// Add events listener
+		if ctrl.listener != nil {
+			ledgerController = ledgercontroller.NewControllerWithEvents(*l, ledgerController, ctrl.listener)
+		}
+
+		return ledgerController, nil
+	})
+}
+
+func (ctrl *DefaultController) CreateLedger(ctx context.Context, name string, configuration ledger.Configuration) error {
+	return tracing.SkipResult(tracing.Trace(ctx, ctrl.tracer, "CreateLedger", tracing.NoResult(func(ctx context.Context) error {
+		configuration.SetDefaults()
+
+		if !ctrl.enableFeatures {
+			if !reflect.DeepEqual(configuration.Features, ledger.DefaultFeatures) {
+				return ErrExperimentalFeaturesDisabled
+			}
+		}
+
+		l, err := ledger.New(name, configuration)
+		if err != nil {
+			return newErrInvalidLedgerConfiguration(err)
+		}
+
+		return ctrl.store.CreateLedger(ctx, l)
+	})))
+}
+
+func (ctrl *DefaultController) GetLedger(ctx context.Context, name string) (*ledger.Ledger, error) {
+	return tracing.Trace(ctx, ctrl.tracer, "GetLedger", func(ctx context.Context) (*ledger.Ledger, error) {
+		return ctrl.store.GetLedger(ctx, name)
+	})
+}
+
+func (ctrl *DefaultController) ListLedgers(ctx context.Context, query ledgercontroller.ListLedgersQuery) (*bunpaginate.Cursor[ledger.Ledger], error) {
+	return tracing.Trace(ctx, ctrl.tracer, "ListLedgers", func(ctx context.Context) (*bunpaginate.Cursor[ledger.Ledger], error) {
+		return ctrl.store.ListLedgers(ctx, query)
+	})
+}
+
+func (ctrl *DefaultController) UpdateLedgerMetadata(ctx context.Context, name string, m map[string]string) error {
+	return tracing.SkipResult(tracing.Trace(ctx, ctrl.tracer, "UpdateLedgerMetadata", tracing.NoResult(func(ctx context.Context) error {
+		return ctrl.store.UpdateLedgerMetadata(ctx, name, m)
+	})))
+}
+
+func (ctrl *DefaultController) DeleteLedgerMetadata(ctx context.Context, param string, key string) error {
+	return tracing.SkipResult(tracing.Trace(ctx, ctrl.tracer, "DeleteLedgerMetadata", tracing.NoResult(func(ctx context.Context) error {
+		return ctrl.store.DeleteLedgerMetadata(ctx, param, key)
+	})))
+}
+
+func NewDefaultController(store Store, listener ledgercontroller.Listener, opts ...Option) *DefaultController {
+	ret := &DefaultController{
+		store:    store,
+		listener: listener,
+		registry: ledgercontroller.NewStateRegistry(),
+		parser:   ledgercontroller.NewDefaultNumscriptParser(),
+	}
+	for _, opt := range append(defaultOptions, opts...) {
+		opt(ret)
+	}
+	return ret
+}
+
+type Option func(ctrl *DefaultController)
+
+func WithUpdateParser(updateParser func(oldParser ledgercontroller.NumscriptParser) ledgercontroller.NumscriptParser) Option {
```

Comment:
You don't need a so "complex" option. Just a `WithUpdateParser(ledgercontroller.NumscriptParser)` feels ok.
And decorating plumber in the module.go.